In [118]:
import yfinance as yf
import random
import json
import pandas as pd


In [119]:
# ISIN für Apple Inc.
isin = "US0378331005"

# Das Tickersymbol für die ISIN abrufen
ticker = yf.Ticker(isin)

# Das Tickersymbol anzeigen
# print(ticker.info['industry'])
print(ticker.info)

{'address1': 'One Apple Park Way', 'city': 'Cupertino', 'state': 'CA', 'zip': '95014', 'country': 'United States', 'phone': '408 996 1010', 'website': 'https://www.apple.com', 'industry': 'Consumer Electronics', 'industryKey': 'consumer-electronics', 'industryDisp': 'Consumer Electronics', 'sector': 'Technology', 'sectorKey': 'technology', 'sectorDisp': 'Technology', 'longBusinessSummary': 'Apple Inc. designs, manufactures, and markets smartphones, personal computers, tablets, wearables, and accessories worldwide. The company offers iPhone, a line of smartphones; Mac, a line of personal computers; iPad, a line of multi-purpose tablets; and wearables, home, and accessories comprising AirPods, Apple TV, Apple Watch, Beats products, and HomePod. It also provides AppleCare support and cloud services; and operates various platforms, including the App Store that allow customers to discover and download applications and digital content, such as books, music, video, games, and podcasts. In add

In [86]:
class Investor:
    # Klassenattribute

    def __init__(self, id:int, name, portfolio, investor_type):
        # Instanzattribute
        self._id = id
        self._name = name
        self._portfolio = portfolio
        self._investor_type = investor_type
        self._bad_stocks = []
        self._bulk_risks = []
        self._market_allocation = 0
        self._sector_allocation = 0
        self._market_allocation_difference = 0
        self._sector_allocation_difference = 0

    # Methoden
    def get_name(self):
        return f"Investor's name is {self._name}."
        
    def calculate_allocation(self):
        pass

    def calculate_allocation_difference(self):
        pass

    def find_bad_stocks(self):
        pass

    def find_bulk_risks(self):
        pass

    def buy(self, isin, amount):
        pass

    def random_portfolio(self):
        pass

    def write_JSON(self):
        pass

    def __str__(self) -> str:
        return f"Investor's name is {self._name}."
    
    def __repr__(self):
        return f"{self._name}"

In [198]:
# Random Investor
investors = []

def random_investor():
    rand_id = ''
    for i in range(6):
        rand_id+=(str(random.randint(0,9)))
    
    name = "John Doe"
    empty_portfolio = []

    investors.append(Investor(rand_id, name, empty_portfolio, "swiss_bias"))

random_investor()
investors


[John Doe]

In [103]:
investors[0]._bad_stocks
investors[0]._id

'443632'

In [113]:
# referenz-allokation laden
profile = 'swiss-bias'
with open(f'..\data\profile_{profile}.json') as json_file:
    reference_allocation = json.load(json_file)
    print("Type:", type(reference_allocation))

Type: <class 'dict'>


In [114]:
print(reference_allocation)

{'name': 'Swiss-Bias Investor', 'markets': {'USA': 0.15, 'Europe': 0.1, 'Switzerland': 0.7, 'Asia & EM': 0.05}, 'sectors': {'Financial Services': 0.18, 'Technology': 0.08, 'Industrials': 0.15, 'Energy': 0.02, 'Healthcare': 0.25, 'Consumer Defensive': 0.15, 'Consumer Cyclical': 0.08, 'Basic Materials': 0.05, 'Real Estate': 0.0, 'Communication Services': 0.02, 'Utilities': 0.02}}


In [115]:
# json-file laden
with open(f'..\data\dummy_portfolio.json') as json_file:
    dummy_portfolio = json.load(json_file)
    print("Type:", type(dummy_portfolio))

Type: <class 'dict'>


In [116]:
print(dummy_portfolio)

{'id': 1, 'name': 'John Doe', 'investor_type': 'Swiss-Bias', 'portfolio': [{'name': 'ABB', 'isin': 'CH0012221716', 'quantity': 10, 'price': 100}, {'name': 'Ralph Lauren', 'isin': 'US7512121010', 'quantity': 15, 'price': 200}], 'bad_stocks': [{'name': 'ABB', 'isin': 'CH0012221716', 'quantity': 10, 'price': 100}, {'name': 'Ralph Lauren', 'isin': 'US7512121010', 'quantity': 15, 'price': 200}], 'market_allocation': {'USA': 0.5, 'Europe': 0.0, 'Switzerland': 0.5, 'Asia & EM': 0.0}, 'sector_allocation': {'Financial Services': 0.01, 'Technology': 0.01, 'Industrials': 0.01, 'Energy': 0.01, 'Healthcare': 0.01, 'Consumer Defensive': 0.01, 'Consumer Cyclical': 0.01, 'Basic Materials': 0.01, 'Real Estate': 0.01, 'Communication Services': 0.01, 'Utilities': 0.01}, 'market_allocation_difference': {'USA': 0.35, 'Europe': -0.1, 'Switzerland': -0.2, 'Asia & EM': -0.05}, 'sector_allocation_difference': {'Financial Services': -0.17, 'Technology': -0.07, 'Industrials': -0.14, 'Energy': -0.01, 'Healthcare'

In [143]:
# stock-info laden
stock_info = pd.read_csv('..\data\stock_info.tsv', sep='\t')
# stock_info['isin'].head()
stock_info['longName'].head()
stock_info['industry'].head(10)

0      Electrical Equipment & Parts
1    Medical Instruments & Supplies
2     Building Products & Equipment
3               Specialty Chemicals
4                Building Materials
5    Integrated Freight & Logistics
6                 Computer Hardware
7            Diagnostics & Research
8                    Packaged Foods
9      Drug Manufacturers - General
Name: industry, dtype: object

In [144]:
print(stock_info.columns)

Index(['address1', 'city', 'zip', 'country', 'phone', 'fax', 'website',
       'industry', 'industryKey', 'industryDisp',
       ...
       'shortPercentOfFloat', 'totalAssets', 'navPrice', 'beta3Year',
       'fundFamily', 'fundInceptionDate', 'legalType', 'yield', 'ytdReturn',
       'threeYearAverageReturn'],
      dtype='object', length=146)


In [185]:
# test-portfolio
portfolio = []
isin_list = ['CH0038863350', 'CH0012032048', 'US4581401001']
quantity_list = [10, 30, 15]

for i in range(len(isin_list)):
    portfolio.append({'isin':isin_list[i], 'quantity':quantity_list[i]})
portfolio

[{'isin': 'CH0038863350', 'quantity': 10},
 {'isin': 'CH0012032048', 'quantity': 30},
 {'isin': 'US4581401001', 'quantity': 15}]

In [197]:
# mit portfolio-positionen die daten aus stock-data holen für allokation
portfolio_value = 0 # variable für gesamtwert des portfolios
for position in portfolio:
    # info = stock_info.loc[stock_info['isin'] == 'isin']
    info = stock_info.loc[stock_info['isin'] == position['isin'], ['isin', 'longName', 'previousClose', 'country' , 'sector']] # zeile aus stock-info von dieser Aktie
    position['name'] = info['longName'].values[0]
    position['price'] = info['previousClose'].values[0]
    position['value'] = position['quantity']*position['price']
    portfolio_value += position['value']    # Portfoliowert aktualisieren
    # position['country'] = info['country'].values[0]
    position['sector'] = info['sector'].values[0]

# Anteil am Portfoliowert für jede Position
for position in portfolio:
    position['portfolio_percent'] = round(position['value']/portfolio_value, 4)
    
print(portfolio_value)
print(portfolio)
portfolio

9157.6
[{'isin': 'CH0038863350', 'quantity': 10, 'name': 'Nestlé S.A.', 'price': 103.42, 'value': 1034.2, 'portfolio_percent': 0.1129, 'sector': 'Consumer Defensive'}, {'isin': 'CH0012032048', 'quantity': 30, 'name': 'Roche Holding AG', 'price': 252.5, 'value': 7575.0, 'portfolio_percent': 0.8272, 'sector': 'Healthcare'}, {'isin': 'US4581401001', 'quantity': 15, 'name': 'Intel Corporation', 'price': 36.56, 'value': 548.4000000000001, 'portfolio_percent': 0.0599, 'sector': 'Technology'}]


[{'isin': 'CH0038863350',
  'quantity': 10,
  'name': 'Nestlé S.A.',
  'price': 103.42,
  'value': 1034.2,
  'portfolio_percent': 0.1129,
  'sector': 'Consumer Defensive'},
 {'isin': 'CH0012032048',
  'quantity': 30,
  'name': 'Roche Holding AG',
  'price': 252.5,
  'value': 7575.0,
  'portfolio_percent': 0.8272,
  'sector': 'Healthcare'},
 {'isin': 'US4581401001',
  'quantity': 15,
  'name': 'Intel Corporation',
  'price': 36.56,
  'value': 548.4000000000001,
  'portfolio_percent': 0.0599,
  'sector': 'Technology'}]

In [169]:
nesn_data = stock_info.loc[stock_info['isin'] == 'CH0038863350', ['isin', 'longName', 'previousClose', 'sector']]
nesn_data


isin     longName  previousClose              sector
8  CH0038863350  Nestlé S.A.         103.42  Consumer Defensive

In [159]:
info = stock_info.loc[stock_info['isin'] == portfolio[0]['isin']]
info

address1   city   zip      country            phone  fax  \
8  avenue NestlE 55  Vevey  1800  Switzerland  41 21 924 11 11  NaN   

                  website        industry     industryKey    industryDisp  \
8  https://www.nestle.com  Packaged Foods  packaged-foods  Packaged Foods   

   ... shortPercentOfFloat totalAssets navPrice beta3Year  fundFamily  \
8  ...                 NaN         NaN      NaN       NaN         NaN   

  fundInceptionDate  legalType  yield  ytdReturn  threeYearAverageReturn  
8               NaN        NaN    NaN        NaN                     NaN  

[1 rows x 146 columns]

Klasse: investor
Attribut:
portfolio isch s dictionary mit beliebig viel key:value paar wo sind ISIN:Anzahl_Aktien (CH0038863350:1400) <-- er het 1400x nestlé aktia
bad_stocks: a lista mit allna aktia wo de CIO uf "SELL" het (warnig für da user, dass er dia söll wegtua)
market_allocation isch s dictionary mit da vier märkt als key und deren %-Anteil als Value (Europe:20%)
sector_allocation: same, aber mit industriesektora (Tech:8%)
investor_type: uswahl vo eina vo da drei muster-allokationa (Swiss-Bias, Swiss-Global, Global)
market_allocation_difference: dictionary wo d unterschied zwüschet de allocation vom investor und de muster-allokation drin isch (investor het 5% USA und s muster het 35% USA -->  USA:-30%)
sector_allocation_difference: same mit sektora
Funktiona:
calculate_allocation: berechnig vo de market_allocation und de sector_allocation anhand vo sinem portfolio
calculate_allocation_differences: berechnig vo market_allocation_difference und sector_allocation_difference
find_bad_stocks: im portfolio d aktia ablicha mit de CIO lista vo "SELL" aktia, und wenn ma eini findet schribt ma sie in bad_stocks lista
buy(ISIN, Anzahl): a aktia ins portfolio kaufa
random_portfolio: portfolio fülla mit ca. 15-25 Aktia, jewiils 15'000 bis 150'000 Franka. ich glaub es söll sich a chli am investor_type usrichta, aber dörf ruhig recht random sii macht eifach das 15-25x:    random aktia ussuacha, random betrag bestimma (15-150k), betrag//(Wert vo 1 Aktia) <-- ganzzahldivision, wia viel stück ma kauft, funktion buy(ISIN, Anzahl) ufrüafa
write_JSON: pack d dictionaries in es JSON (fürs frontend) (bearbeitet)